encoder.go

```diff
@@ -320,3 +373,111 @@ func (encoder *encoder) encodeArray(value reflect.Value, obj *wafObject, depth i
 	// Fix the size because we skipped map entries
 	obj.nbEntries = uint64(currIndex)
 }
+
+func (encoder *encoder) addTruncation(reason TruncationReason, size int) {
+	if encoder.truncations == nil {
+		encoder.truncations = make(map[TruncationReason][]int, 3)
+	}
+	encoder.truncations[reason] = append(encoder.truncations[reason], size)
+}
+
+// mesureObjectDepth traverses the provided object recursively to try and obtain
+// the real object depth, but limits itself to about 1ms of time budget, past
+// which it'll stop and return whatever it has go to so far.
+func (encoder *encoder) measureObjectDepth(obj reflect.Value) {
+	ctx, cancelCtx := context.WithTimeout(context.Background(), time.Millisecond)
+	defer cancelCtx()
+
+	depth, _ := depthOf(ctx, obj)
+	encoder.truncations[ObjectTooDeep] = []int{depth}
+}
+
+// depthOf returns the depth of the provided object. This is 0 for scalar values,
+// such as strings.
+func depthOf(ctx context.Context, obj reflect.Value) (int, error) {
+	if err := ctx.Err(); err != nil {
+		// Timed out, won't go any deeper
+		return 0, err
+	}
+
+	obj, kind := resolvePointer(obj)
+
+	switch kind {
+	case reflect.Array, reflect.Slice:
+		if obj.Type() == reflect.TypeOf([]byte(nil)) {
+			// We treat byte slices as strings
+			return 0, nil
+		}
+		maxItemDepth := 0
+		var (
+			itemDepth int
+			err       error
+		)
+		for i := 0; i < obj.Len(); i++ {
+			itemDepth, err = depthOf(ctx, obj.Index(i))
+			if itemDepth > maxItemDepth {
+				maxItemDepth = itemDepth
+			}
+			if err != nil {
+				break
+			}
+		}
+		return maxItemDepth + 1, err
+	case reflect.Map:
+		maxValueDepth := 0
+		var (
+			valueDepth int
+			err        error
+		)
+		for iter := obj.MapRange(); iter.Next(); {
+			valueDepth, err = depthOf(ctx, iter.Value())
+			if valueDepth > maxValueDepth {
+				maxValueDepth = valueDepth
+			}
+			if err != nil {
+				break
+			}
+		}
+		return maxValueDepth + 1, err
+	case reflect.Struct:
+		maxFieldDepth := 0
+		typ := obj.Type()
+		var (
+			fieldDepth int
+			err        error
+		)
```

Comment:
Can we put atop of the switch all these variables in common for all switch cases?